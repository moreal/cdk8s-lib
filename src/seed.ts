// import { Deployment, Protocol } from "cdk8s-plus-22";
import { Construct } from "constructs";
import { KubeDeployment, KubeService, IntOrString, Quantity } from "cdk8s-plus-22/lib/imports/k8s";

interface HostAndPort { host: string, port: number };

type LogLevel = "error" | "warning" | "information" | "debug" | "verbose"

export interface SeedOptions {
    image: string,
    logLevel: LogLevel,
    workers: number | undefined,
    appProtocolVersion: string,
    libplanet: HostAndPort,
    graphql: HostAndPort | undefined,
    privateKey: string,
};

export class Seed extends Construct {
    constructor(scope: Construct, id: string, options: SeedOptions) {
        super(scope, id);

        new KubeDeployment(this, `${id}-deployment`, {
            metadata: {
                labels: {
                    app: id
                },
                name: id,
            },
            spec: {
                replicas: 1,
                selector: {
                    matchLabels: {
                        app: id,
                    }
                },
                template: {
                    spec: {
                        containers: [
                            {
                                name: id,
                                image: options.image,
                                command: ["dotnet"],
                                args: [
                                    "Libplanet.Seed.Executable.dll",
                                    "run",
                                    `--log-level=${options.logLevel}`,
                                    `--app-protocol-version=${options.appProtocolVersion}`,
                                    `--host=${options.libplanet.host}`,
                                    `--port=${options.libplanet.port}`,
                                    ...(options.workers === undefined ? [] : [`--workers=${options.workers}`]),
                                    `--private-key=${options.privateKey}`,
                                    ...(options.graphql === undefined
                                        ? []
                                        : [
                                            `--graphql-host=${options.graphql.host}`,
                                            `--graphql-port=${options.graphql.port}`
                                        ]),
                                ],
                                ports: [
                                    {
                                        containerPort: options.libplanet.port,
                                        name: "node",
                                        protocol: "TCP",
                                    },
                                    ...(options.graphql !== undefined ?
                                        [{
                                            containerPort: options.graphql.port,
                                            name: "graphql",
                                            protocol: "TCP",
                                        }] : [])
                                ],
                                livenessProbe: {
                                    failureThreshold: 3,
                                    initialDelaySeconds: 120,
                                    periodSeconds: 5,
                                    successThreshold: 1,
                                    tcpSocket: {
                                        port: IntOrString.fromNumber(options.libplanet.port),
                                    },
                                    timeoutSeconds: 1,
                                },
                                resources: {
                                    requests: {
                                        cpu: Quantity.fromString(String(1)),
                                    }
                                }
                            }
                        ],
                        dnsPolicy: "ClusterFirst",
                    }
                }
            }
        });

        new KubeService(this, `${id}-service`, {
            metadata: {
                name: id,
            },
            spec: {
                selector: {
                    app: id,
                },
                type: "LoadBalancer",
                ports: [
                    {
                        name: "node",
                        port: options.libplanet.port,
                        targetPort: IntOrString.fromNumber(options.libplanet.port),
                    },
                    ...(options.graphql === undefined
                        ? []
                        : [{
                            name: "graphql",
                            port: options.graphql.port,
                            targetPort: IntOrString.fromNumber(options.graphql.port),
                        }]),
                ]
            }
        });
    }
}
