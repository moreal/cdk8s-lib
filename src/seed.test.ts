import { Seed } from './seed';
import { Testing } from 'cdk8s';

describe(Seed.name, () => {
    test('snapshot', () => {
        const chart = Testing.chart();
        new Seed(chart, 'main-seed-1', {
            image: "planetariumhq/libplanet-seed:git-08e65a8b4a88b9c53266a92df573641a7b6e50f0",
            appProtocolVersion: "100080/6ec8E598962F1f475504F82fD5bF3410eAE58B9B/MEUCIQCrSQ9v7NZBo5yCyvlHgBXW6lD4p2lFlKqK+aMeoKA.IAIgVkcK6hMCw1E0mf+m7rCf+9XnSL8Apg.joe6I1a8Nw70=/ZHUxNjpXaW5kb3dzQmluYXJ5VXJsdTU2Omh0dHBzOi8vZG93bmxvYWQubmluZS1jaHJvbmljbGVzLmNvbS92MTAwMDgwL1dpbmRvd3MuemlwdTk6dGltZXN0YW1wdTEwOjIwMjEtMDktMzBl",
            graphql: {
                host: "0.0.0.0",
                port: 31237,
            },
            libplanet: {
                host: "9c-main-seed-1.planetarium.dev",
                port: 31234,
            },
            logLevel: "debug",
            // Created from `planet key generate`.
            privateKey: "09aa7ce7bbd71d50bcf34e089545f1875bbfbc7bea56a24467bc9e7e79228d3e",
            workers: 500,
        });
        const results = Testing.synth(chart)
        expect(results).toMatchSnapshot();
    });
});
