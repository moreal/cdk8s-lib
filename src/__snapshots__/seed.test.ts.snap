// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seed snapshot 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app": "main-seed-1",
      },
      "name": "main-seed-1",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "main-seed-1",
        },
      },
      "template": Object {
        "spec": Object {
          "containers": Array [
            Object {
              "args": Array [
                "Libplanet.Seed.Executable.dll",
                "run",
                "--log-level=debug",
                "--app-protocol-version=100080/6ec8E598962F1f475504F82fD5bF3410eAE58B9B/MEUCIQCrSQ9v7NZBo5yCyvlHgBXW6lD4p2lFlKqK+aMeoKA.IAIgVkcK6hMCw1E0mf+m7rCf+9XnSL8Apg.joe6I1a8Nw70=/ZHUxNjpXaW5kb3dzQmluYXJ5VXJsdTU2Omh0dHBzOi8vZG93bmxvYWQubmluZS1jaHJvbmljbGVzLmNvbS92MTAwMDgwL1dpbmRvd3MuemlwdTk6dGltZXN0YW1wdTEwOjIwMjEtMDktMzBl",
                "--host=9c-main-seed-1.planetarium.dev",
                "--port=31234",
                "--workers=500",
                "--private-key=09aa7ce7bbd71d50bcf34e089545f1875bbfbc7bea56a24467bc9e7e79228d3e",
                "--graphql-host=0.0.0.0",
                "--graphql-port=31237",
              ],
              "command": Array [
                "dotnet",
              ],
              "image": "planetariumhq/libplanet-seed:git-08e65a8b4a88b9c53266a92df573641a7b6e50f0",
              "livenessProbe": Object {
                "failureThreshold": 3,
                "initialDelaySeconds": 120,
                "periodSeconds": 5,
                "successThreshold": 1,
                "tcpSocket": Object {
                  "port": 31234,
                },
                "timeoutSeconds": 1,
              },
              "name": "main-seed-1",
              "ports": Array [
                Object {
                  "containerPort": 31234,
                  "name": "node",
                  "protocol": "TCP",
                },
                Object {
                  "containerPort": 31237,
                  "name": "graphql",
                  "protocol": "TCP",
                },
              ],
              "resources": Object {
                "requests": Object {
                  "cpu": "1",
                },
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "main-seed-1",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "node",
          "port": 31234,
          "targetPort": 31234,
        },
        Object {
          "name": "graphql",
          "port": 31237,
          "targetPort": 31237,
        },
      ],
      "selector": Object {
        "app": "main-seed-1",
      },
      "type": "LoadBalancer",
    },
  },
]
`;
